### ISSUE TITLE
Skorch incompatibility, eeg augmentation 

### ISSUE THREAD

**Issue by bruAristimunha** [braindecode/braindecode#382]
This error is a little confusing to explain, but I will try my best to explain it. 

Unfortunately, I need some help before submitting a pull request. I was trying to perform a parameter search with GridCVSearch following the methodology of the article developed by @cedricrommel and @jpaillard, available on his [page](https://cedricrommel.github.io/assets/pdfs/systematic_comparison.pdf).

I started with the data augmentation tutorial available in braindecode and used what is contained in the Pull Request number #349 to implement a parameter search using the skorch infrastructure. 

The first error that appeared is associated with the recently implemented transformation functions in braindecode and related to Skorch compatibility (commit #361).

To the best of our knowledge, for skorch to work correctly, all submodules inside the EEGClassifier must be able to receive a parameter (deep=True), or some kwargs, or inherit an estimator from sklearn. 

I consulted the skorch team [on this issue](https://github.com/skorch-dev/skorch/issues/794), and I debugged this error in more detail, and I arrived that the core issue is contained in the clone part: 

![image](https://user-images.githubusercontent.com/42702466/167723938-adff495a-c6d9-4a61-b329-e345c46a061f.png)

Following this, I tried to add a deep parameter to one of the transformations, such as the first picture below, and I got a different error message. Still, there is a new problem with redefining the augmentation functions.

![image](https://user-images.githubusercontent.com/42702466/167723874-f2135770-3377-41bd-90f3-bb98c0d0542e.png)

![image](https://user-images.githubusercontent.com/42702466/167723829-e41998ac-badb-4d10-b3dd-9954cb935a56.png)

Given this scenario, I believe there is some minor bug in skorch compatibility with the augmentation functions.

It is possible to work around this issue by performing cross-validation by hand and without the sklearn library. So I don't know how relevant this issue is.

**Comment by agramfort** [comment]
can you gist a small code snippet to replicate the error? first thing is to
see if we can replicate on our machines.

**Comment by bruAristimunha** [comment]
``python
import numpy as np

import torch 
import mne

from sklearn.model_selection import KFold, GridSearchCV

from skorch.helper import SliceDataset
from skorch.callbacks import LRScheduler

from braindecode import EEGClassifier
from braindecode.datasets import MOABBDataset
from braindecode.preprocessing import (
    exponential_moving_standardize, preprocess, Preprocessor, scale)
from braindecode.preprocessing import create_windows_from_events
from braindecode.augmentation import GaussianNoise, AugmentedDataLoader
from braindecode.util import set_random_seeds
from braindecode.models import ShallowFBCSPNet

mne.set_log_level('WARNING')

subject_id = 3
seed = 20200220
lr = 0.0625 * 0.01
weight_decay = 0
batch_size = 64
n_epochs = 4
low_cut_hz = 4.  # low cut frequency for filtering
high_cut_hz = 38.  # high cut frequency for filtering
factor_new = 1e-3
init_block_size = 1000
trial_start_offset_seconds = -0.5

#--------------------------------------------------------------------------
# Reading part with pre-processing, copy from braindecode;

dataset = MOABBDataset(dataset_name="BNCI2014001", subject_ids=[subject_id])

preprocessors = [
    Preprocessor('pick_types', eeg=True, meg=False, stim=False),  # Keep EEG sensors
    Preprocessor(scale, factor=1e6, apply_on_array=True),  # Convert from V to uV
    Preprocessor('filter', l_freq=low_cut_hz, h_freq=high_cut_hz),  # Bandpass filter
    Preprocessor(exponential_moving_standardize,  # Exponential moving standardization
                 factor_new=factor_new, init_block_size=init_block_size)
]

preprocess(dataset, preprocessors)

# Extract sampling frequency, check that they are same in all datasets
sfreq = dataset.datasets[0].raw.info['sfreq']
assert all([ds.raw.info['sfreq'] == sfreq for ds in dataset.datasets])
# Calculate the trial start offset in samples.
trial_start_offset_samples = int(trial_start_offset_seconds * sfreq)

# Create windows using braindecode function for this. It needs parameters to
# define how trials should be used.
windows_dataset = create_windows_from_events(
    dataset,
    trial_start_offset_samples=trial_start_offset_samples,
    trial_stop_offset_samples=0,
    preload=True,
)

splitted = windows_dataset.split('session')
train_set = splitted['session_T']
valid_set = splitted['session_E']

## Copy from gemeinl branch

train_X = SliceDataset(train_set, idx=0)
train_y = np.array([y for y in SliceDataset(train_set, idx=1)])


#--------------------------------------------------------------------------
# Model part


cuda = torch.cuda.is_available()  # check if GPU is available, if True chooses to use it
device = 'cuda' if cuda else 'cpu'
if cuda:
    torch.backends.cudnn.benchmark = True

set_random_seeds(seed=seed, cuda=cuda)

n_classes = 4

# Extract number of chans and time steps from dataset
n_channels = train_set[0][0].shape[0]
input_window_samples = train_set[0][0].shape[1]

model = ShallowFBCSPNet(
    n_channels,
    n_classes,
    input_window_samples=input_window_samples,
    final_conv_length='auto',
)

# Send model to GPU
if cuda:
    model.cuda()
    


clf = EEGClassifier(
    model,
    iterator_train=AugmentedDataLoader,  # This tells EEGClassifier to use a custom DataLoader
    iterator_train__transforms=[],  # This sets the augmentations to use. pass in CV search
    criterion=torch.nn.NLLLoss,
    optimizer=torch.optim.AdamW,
    train_split=None,  # using valid_set for validation
    optimizer__lr=lr,
    optimizer__weight_decay=weight_decay,
    batch_size=batch_size,
    callbacks=[
        "accuracy",
        ("lr_scheduler", LRScheduler('CosineAnnealingLR', T_max=n_epochs - 1)),
    ],
    device=device,
)



#--------------------------------------------------------------------------
# Cross validation part


cv = KFold(n_splits=3, shuffle=True, random_state=42)
fit_params = {'epochs': n_epochs}


transforms = [GaussianNoise(probability=1.,  std=0.1, random_state=2), 
             GaussianNoise(probability=1.,  std=0.5, random_state=2 )] 

param_grid = {
    'iterator_train__transforms': transforms,
}
search = GridSearchCV(
    estimator=clf,
    param_grid=param_grid,
    cv=cv,
    return_train_score=True,
    scoring='accuracy',
    refit=True,
    verbose=1,
    error_score='raise')

# Error:

search.fit(train_X, train_y, **fit_params)
``

**Comment by bruAristimunha** [comment]
Dear @agramfort, below the code to replicate the error:

```python
import numpy as np

import torch 
import mne

from sklearn.model_selection import KFold, GridSearchCV

from skorch.helper import SliceDataset
from skorch.callbacks import LRScheduler

from braindecode import EEGClassifier
from braindecode.datasets import MOABBDataset
from braindecode.preprocessing import (
    exponential_moving_standardize, preprocess, Preprocessor, scale)
from braindecode.preprocessing import create_windows_from_events
from braindecode.augmentation import GaussianNoise, AugmentedDataLoader
from braindecode.util import set_random_seeds
from braindecode.models import ShallowFBCSPNet

mne.set_log_level('WARNING')

subject_id = 3
seed = 20200220
lr = 0.0625 * 0.01
weight_decay = 0
batch_size = 64
n_epochs = 4
low_cut_hz = 4.  # low cut frequency for filtering
high_cut_hz = 38.  # high cut frequency for filtering
factor_new = 1e-3
init_block_size = 1000
trial_start_offset_seconds = -0.5

#--------------------------------------------------------------------------
# Reading part with pre-processing, copy from braindecode;

dataset = MOABBDataset(dataset_name="BNCI2014001", subject_ids=[subject_id])

preprocessors = [
    Preprocessor('pick_types', eeg=True, meg=False, stim=False),  # Keep EEG sensors
    Preprocessor(scale, factor=1e6, apply_on_array=True),  # Convert from V to uV
    Preprocessor('filter', l_freq=low_cut_hz, h_freq=high_cut_hz),  # Bandpass filter
    Preprocessor(exponential_moving_standardize,  # Exponential moving standardization
                 factor_new=factor_new, init_block_size=init_block_size)
]

preprocess(dataset, preprocessors)

# Extract sampling frequency, check that they are same in all datasets
sfreq = dataset.datasets[0].raw.info['sfreq']
assert all([ds.raw.info['sfreq'] == sfreq for ds in dataset.datasets])
# Calculate the trial start offset in samples.
trial_start_offset_samples = int(trial_start_offset_seconds * sfreq)

# Create windows using braindecode function for this. It needs parameters to
# define how trials should be used.
windows_dataset = create_windows_from_events(
    dataset,
    trial_start_offset_samples=trial_start_offset_samples,
    trial_stop_offset_samples=0,
    preload=True,
)

splitted = windows_dataset.split('session')
train_set = splitted['session_T']
valid_set = splitted['session_E']

## Copy from gemeinl branch

train_X = SliceDataset(train_set, idx=0)
train_y = np.array([y for y in SliceDataset(train_set, idx=1)])


#--------------------------------------------------------------------------
# Model part


cuda = torch.cuda.is_available()  # check if GPU is available, if True chooses to use it
device = 'cuda' if cuda else 'cpu'
if cuda:
    torch.backends.cudnn.benchmark = True

set_random_seeds(seed=seed, cuda=cuda)

n_classes = 4

# Extract number of chans and time steps from dataset
n_channels = train_set[0][0].shape[0]
input_window_samples = train_set[0][0].shape[1]

model = ShallowFBCSPNet(
    n_channels,
    n_classes,
    input_window_samples=input_window_samples,
    final_conv_length='auto',
)

# Send model to GPU
if cuda:
    model.cuda()
    


clf = EEGClassifier(
    model,
    iterator_train=AugmentedDataLoader,  # This tells EEGClassifier to use a custom DataLoader
    iterator_train__transforms=[],  # This sets the augmentations to use. pass in CV search
    criterion=torch.nn.NLLLoss,
    optimizer=torch.optim.AdamW,
    train_split=None,  # using valid_set for validation
    optimizer__lr=lr,
    optimizer__weight_decay=weight_decay,
    batch_size=batch_size,
    callbacks=[
        "accuracy",
        ("lr_scheduler", LRScheduler('CosineAnnealingLR', T_max=n_epochs - 1)),
    ],
    device=device,
)



#--------------------------------------------------------------------------
# Cross validation part


cv = KFold(n_splits=3, shuffle=True, random_state=42)
fit_params = {'epochs': n_epochs}


transforms = [GaussianNoise(probability=1.,  std=0.1, random_state=2), 
             GaussianNoise(probability=1.,  std=0.5, random_state=2 )] 

param_grid = {
    'iterator_train__transforms': transforms,
}
search = GridSearchCV(
    estimator=clf,
    param_grid=param_grid,
    cv=cv,
    return_train_score=True,
    scoring='accuracy',
    refit=True,
    verbose=1,
    error_score='raise')

# Error:

search.fit(train_X, train_y, **fit_params)
```
When a debug the error and find that the core error is related to this sequence of code:

```python
from sklearn.base import clone

clone(transforms[0])
```
Basically, when we pass the estimator, it tries to copy all the modules and sub-modules, and the class of transformation in braindecode doesn't have this parameter named `deep`.

a link for a colab: https://colab.research.google.com/drive/1veQPVo3_sElFYXFS0A72YZU_DCkKkplk?usp=sharing

**Comment by bruAristimunha** [comment]
Amazing @agramfort! 

If you and your teams support this suggestion, I can refactor the code to rename the function the get_params and change it for something else.

I think I would need a new name.

**Comment by agramfort** [comment]
@cedricrommel @gemeinl @robintibor do you have any opinion here?

**Comment by gemeinl** [comment]
Thx @bruAristimunha and @agramfort . I was not deeply involved in the augmentation code, but renaming get_params sounds fine. If this ensures compatibility with scikit-learn I am happy. What do you think @cedricrommel ?

**Comment by cedricrommel** [comment]
Thanks a lot @bruAristimunha and @agramfort ! `get_params` was initially named like this to reproduce torchvision API. But I see no problem in renaming it if it helps ensuring scikit-learn compatibility :)

**Comment by agramfort** [comment]
@cedricrommel any suggestion on the new name? Shall we make get_params return all params with probability and then create a get_magnitude_params ?

**Comment by cedricrommel** [comment]
Why not `get_functional_params` or `get_operation_params` ? Not a fan of `get_magnitude°params` because these functions return many arguments which do not all reflect the magnitude of the transform.

**Comment by gemeinl** [comment]
How about `get_augmentation_params`?

**Comment by agramfort** [comment]
>
> How about get_augmentation_params?Message ID:
> ***@***.***>
>
I like that !

**Comment by agramfort** [comment]
it will get closed automatically when the PR is merged.
